### ISSUE TITLE
C# new preprocessor directives "#pragma exception before"

### ISSUE THREAD

**Issue by srburton** [dotnet/roslyn#30641]
**2.1.403**: 

**Steps to Reproduce**:

` New implementation`

**Expected Behavior**:

`#pragma  exeception before`

This is the normal situation that we all know today, I did a treatment with an alert ja and when I give an exception within a delegate that execption of a warning in Visual Studio the #pragma would be to disable this warning that Visual Studio and continue the Code because it is an exception that has already been handled.

![47243054-3a029300-d3c7-11e8-9264-13c157a09ac0](https://user-images.githubusercontent.com/15394508/47290027-ee710480-d5d3-11e8-883d-8a3aa89322bf.jpg)

**Sugestion:**

![47243160-98c80c80-d3c7-11e8-9dee-b820527722b4](https://user-images.githubusercontent.com/15394508/47290056-06e11f00-d5d4-11e8-910d-304dafcf6dee.jpg)

**Reference:**

https://github.com/dotnet/csharplang/issues/1936

**Comment by svick** [comment]
Why aren't the existing options offered by VS sufficient for you?

Also, does the behavior improve if you use `foreach` instead of `ForEach`?

**Comment by srburton** [comment]
Example:

![47242692-ef344b80-d3c5-11e8-8ef2-8ba3118125a0](https://user-images.githubusercontent.com/15394508/47294353-33e8fe00-d5e3-11e8-9aaf-825e9e140e47.jpg)

**Comment by CyrusNajmabadi** [comment]
This issue lacks enogh information to be actionable.

**Comment by srburton** [comment]
@CyrusNajmabadi 
Do you have any models that I can follow to publish something more complete here?

**Comment by srburton** [comment]
@svick 
no, because with settings in the VS gets on the machine on the user and when shared the project will start to give numerous errors on the machine of another programmer, with these #pragma gets more declarative that there has an exception that burst in the first try/cat H Previous, is functionality would be very useful for lambdas.

**Comment by CyrusNajmabadi** [comment]
> No, because with settings in the VS gets on the machine on the user and when shared the project will start to give numerous errors on the machine of another programmer,

Why is that a problem?  How debuggers work is entirely up to the individual.  If i choose to use windbg, i want to have control over that experience.  If someone else wants to use a debugger present in their IDE, that is up to them.

**Comment by CyrusNajmabadi** [comment]
> would be to disable this warning that Visual Studio and continue the Code because it is an exception that has already been handled.

You can just use a boolean for this:

```c#
try 
{
    // ...
}
catch (Exception e)
{
    if (!shown)
    {
        shown = true;
        _alert().Show(e.Message);
    }
}
```

Now you'll only ever report that exception handler once.

**Comment by srburton** [comment]
@CyrusNajmabadi 

Good if you want to do with activators like Boolean that gets a dirty and great code all right just got another point of view that would be very useful a #pragma to disable exception of a specific snippet of my code, do not want to disable my entire project.

**Comment by CyrusNajmabadi** [comment]
> Good if you want to do with activators like Boolean 

I'm having trouble understanding what you're trying to say here.

It sounds like you think this is a good solution. If so... great!  It doesn't look like we need language changes for this.

> that would be very useful a #pragma to disable exception of a specific snippet of my code

What does it mean to "disable exception"?  Do you mean: the exception won't be thrown?  Or do you mean: the debugger won't report the exception?

**Comment by srburton** [comment]
I'm using a translation tool, the debugger will not report the exception in that specific code block.

**Comment by CyrusNajmabadi** [comment]
> I'm using a translation tool, the debugger will not report the exception in that specific code block.

**Comment by srburton** [comment]
the **debugger** will **not report the exception** in that **specific code block**.

**Comment by CyrusNajmabadi** [comment]
Ok.  So it sounds like this is purely a tooling request, and is unrelated to making changes to the language or adding pragmas.  

For example, WinDbg gives a lot of control over exceptions:
https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/controlling-exceptions-and-events#span-idcontrollingexceptionsandeventsfromthedebuggerspanspan-idcontrollingexceptionsandeventsfromthedebuggerspancontrolling-exceptions-and-events-from-the-debugger

If you'd like the VS debugger to have support for some sort of "don't actually pause for exceptions within this block" i recommend you file a VSFeedback item here:

![image](https://user-images.githubusercontent.com/4564579/47380693-dfad4f00-d6b2-11e8-9c04-9d797027ddf1.png)

**Comment by CyrusNajmabadi** [comment]
Closing out as this is a debugger request, not a roslyn one.
